### ISSUE TITLE
Powershell commands not utilizing proxy settings

### ISSUE THREAD

**Issue by Zero5211** [aws/amazon-ssm-agent#65]
I've been struggling with this issue for the past week and finally found a solution. I have the proxy variables configured under "HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Services\AmazonSSMAgent\Environment", but any PowerShell commands run by SSM would not use the proxy. Setting the proxy in IE, and using netsh winhttp set proxy were not causing it to use the proxy either. Since SSM runs as a service under Local System, it uses "NT Authority\System" to run commands. However, according to [this](https://serverfault.com/questions/34940/how-do-i-configure-proxy-settings-for-local-system) that user does not use the same proxy setting source as others. I followed the instructions there and exported the Connections key from a user with the proxy configured and then imported it to the S-1-5-18 user, and it finally used the proxy.

Here's an example. I have my proxy configured to block www.google.com. First, without that key in the registry, using AWSRunPowerShellScript and input "netsh winhttp show proxy; wget www.google.com":

```
Current WinHTTP proxy settings:

    Proxy Server(s) :  10.27.133.163:3128
    Bypass List     :  169.254.169.254



StatusCode        : 200
StatusDescription : OK
Content           : <!doctype html><html itemscope="" 
                    itemtype="http://schema.org/WebPage" lang="en"><head><meta 
                    content="Search the world's information, including 
                    webpages, images, videos and more. Google has many speci...
RawContent        : HTTP/1.1 200 OK
                    X-XSS-Protection: 1; mode=block
                    X-Frame-Options: SAMEORIGIN
                    Vary: Accept-Encoding
                    Transfer-Encoding: chunked
                    Accept-Ranges: none
                    Cache-Control: private, max-age=0
                    Content-Type: ...
Forms             : {f}
Headers           : {[X-XSS-Protection, 1; mode=block], [X-Frame-Options, 
                    SAMEORIGIN], [Vary, Accept-Encoding], [Transfer-Encoding, 
                    chunked]...}
Images            : {@{innerHTML=; innerText=; outerHTML=<IMG id=hplogo 
                    onload=window.lol&amp;&amp;lol() style="PADDING-BOTTOM: 
                    14px; PADDING-TOP: 28px; PADDING-LEFT: 0px; PADDING-RIGHT: 
                    0px" alt=Google src="/images/branding/googlelogo/1x/googlel
                    ogo_white_background_color_272x92dp.png" width=272 
                    height=92>; outerText=; tagName=IMG; id=hplogo; 
                    onload=window.lol&amp;&amp;lol(); style=PADDING-BOTTOM: 
                    14px; PADDING-TOP: 28px; PADDING-LEFT: 0px; PADDING-RIGHT: 
                    0px; alt=Google; src=/images/branding/googlelogo/1x/googlel
                    ogo_white_background_color_272x92dp.png; width=272; 
                    height=92}}
InputFields       : {@{innerHTML=; innerText=; outerHTML=<INPUT type=hidden 
                    value=en name=hl>; outerText=; tagName=INPUT; type=hidden; 
                    value=en; name=hl}, @{innerHTML=; innerText=; 
                    outerHTML=<INPUT type=hidden value=hp name=source>; 
                    outerText=; tagName=INPUT; type=hidden; value=hp; 
                    name=source}, @{innerHTML=; innerText=; outerHTML=<INPUT 
                    type=hidden name=biw>; outerText=; tagName=INPUT; 
   
---Output truncated---
```

And here is the same after adding that key in:

```
Current WinHTTP proxy settings:

    Proxy Server(s) :  10.27.133.163:3128
    Bypass List     :  169.254.169.254


----------ERROR-------
wget : ERROR
The requested URL could not be retrieved
The following error was encountered while trying to retrieve the URL: 
http://www.google.com/
Access Denied.
Access control configuration prevents your request from being allowed at this 
time. Please contact your service provider if you feel this is incorrect.
Your cache administrator is webmaster.
Generated Fri, 18 Aug 2017 20:53:16 GMT by pocLinuxTC (squid/3.3.8)
At C:\ProgramData\Amazon\SSM\InstanceData\i-08d3fcfa25c2b6233\document\orchestr
ation\c0372a21-0c92-4690-856f-a138b6f13658\awsrunPowerShellScript\0.awsrunPower
ShellScript\_script.ps1:1 char:27
+ netsh winhttp show proxy; wget www.google.com
+                           ~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (System.Net.HttpWebRequest:Htt 
   pWebRequest) [Invoke-WebRequest], WebException
    + FullyQualifiedErrorId : WebCmdletWebResponseException,Microsoft.PowerShe 
   ll.Commands.InvokeWebRequestCommand
```

I can probably get around this issue in my own environment by just importing that key on all of the servers I want to run SSM on, but I was wondering if there's a way to have SSM set this itself. [This thread explores that in more detail.](http://www.vbforums.com/showthread.php?615595-RESOLVED-Editing-the-DefaultConnectionSettings-key-for-proxy-settings-in-ie) There may be another way to get this fixed, but this is all I've been able to find so far.

**Comment by JamieICG** [comment]
Hi, 

This may or may not be related, but I had an issue with installing and configuring the SSM agent behind a proxy.  The problem was that the HTTP proxy variables were not being passed to the SSM installer, and was then failing.  This was accepted as a bug by AWS for fix in a future release.  This was on the 18th of August.

**Comment by mmendonca3** [comment]
Hello Zero5211, 

Proxy issue in SSM Agent was fixed in version 2.0.922.0 (https://github.com/aws/amazon-ssm-agent/tree/2.0.922.0). I will be closing this issue. Please feel free to reopen it if you have any further concerns.  Thank you.

**Comment by j2clerck** [comment]
Hello,

It seems that it isn't resolved as of today. I just deployed a Windows instance behind a proxy and I still had to manually copy the Connections key to S-1-5-18 (aka NT Authority\System) to get it working.

Regards,
Joseph

**Comment by luke-clapham** [comment]
I used the latest version of the Amazon SSM Agent - Release 2.3.13.0- 2018-08-16 - does not work.
Still requires the workaround to import a already existing reg key into - 
HKEY_USERS\S-1-5-18
From
Software\Microsoft\Windows\CurrentVersion\Internet Settings\Connections\DefaultConnectionSettings

**Comment by coofercat** [comment]
I found this thread while trying to get amazon-ssm-agent - v2.3.344.0 to join a domain behind a proxy (during Userdata execution).

I've tried the DefaultConnectionSettings workaround, and had limited success (some SSM functions seem to proxy, but others don't). SSM also seems to report, on start up, suitable proxy configuration but then fails to work. WinHTTP settings don't help either, it seems.

I found https://docs.aws.amazon.com/systems-manager/latest/userguide/sysman-install-ssm-proxy.html which describes yet another set of Registry keys to mess with - and seems to work. With this in place, I didn't need the workaround. My system seems to join the AD domain and reports to AWS (once suitable instance permissions were in place).
